File: 2023/day02/solution.py
#!/usr/bin/env python3
import sys

DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True
D4 = [(-1, 0), (0, 1), (1, 0), (0, -1)]

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i':
        FILENAME = 'input.txt'
      case 'p1':
        PART1, PART2 = True, False
      case 'p2':
        PART1, PART2 = False, True
      case 'p0':
        PART1, PART2 = False, False
      case _:
        raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  text = f.read()
games = [
  {
    'id': int(split0[0].split(' ')[1]),
    'reveals':[
      {
        vk1[1]:int(vk1[0])
        for vk0 in reveal.split(', ')
        if (vk1:=vk0.split(' '))
      }
      for reveal in split0[1].split('; ')
    ]
  }
  for x in text.splitlines() \
  if (split0 := x.split(': ')) \
]


def part1():
  total = 0
  for game in games:
    possible = all(x.get('red', 0) <= 12 for x in game['reveals']) and \
               all(x.get('green', 0) <= 13 for x in game['reveals']) and \
               all(x.get('blue', 0) <= 14 for x in game['reveals']) #type: ignore
    if DEBUG > 0: print(f'{game=} {possible=}')
    if possible: total += game['id']  #type: ignore
  print(f'part 1: {total}')


def part2():
  total = 0
  for game in games:
    red = max(x.get('red', 0) for x in game['reveals'])  #type: ignore
    green = max(x.get('green', 0) for x in game['reveals'])  #type: ignore
    blue = max(x.get('blue', 0) for x in game['reveals'])  #type: ignore
    if DEBUG > 0: print(f'{game=} {red=} {green=} {blue=}')
    total += red * green * blue
  print(f'part 2: {total}')


if PART1: part1()
if PART2: part2()

File: 2023/day02/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = Deno.args.reduce((acc, item) => (item.startsWith('-d') ? Number(item.slice(2) || '1') : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) => (item == '-i' ? 'input.txt' : item.startsWith('-e') ? `example${item.slice(2)}.txt` : acc),
  'example.txt'
);
const [PART1, PART2] = Deno.args.reduce(
  (acc, item) => (item == '-p0' ? [false, false] : item == '-p1' ? [true, false] : item == '-p2' ? [false, true] : acc),
  [true, true]
);

console.log({ FILENAME, DEBUG, PART1, PART2 });

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (DEBUG <= level) console.debug(...data);
};

const input = await Deno.readTextFile(FILENAME);

interface Dice {
  red: number;
  green: number;
  blue: number;
}

const parseInput = () =>
  input
    .replaceAll(';', ',')
    .split('\n')
    .filter((line) => line.trim())
    .map((line) => {
      const parts = line.split(': ');
      return {
        id: Number(parts[0].substring(5)),
        dice: parts[1]
          .split(', ')
          .map((token) => token.split(' '))
          .reduce(
            (acc, [count, colour]) => {
              const index = colour as keyof Dice;
              acc[index] = Math.max(acc[index], Number(count));
              return acc;
            },
            { red: 0, green: 0, blue: 0 } as Dice
          ),
      };
    });

const part1 = () => {
  const games = parseInput();
  debug(1, games);
  const minimums = { red: 12, green: 13, blue: 14 } as Dice;
  const possible = games.reduce(
    (acc, game) =>
      game.dice.red <= minimums.red && game.dice.green <= minimums.green && game.dice.blue <= minimums.blue ? acc + game.id : acc,
    0
  );
  console.log('part 1:', possible);
};

const part2 = () => {
  const games = parseInput();
  const power = games.reduce((acc, game) => acc + game.dice.red * game.dice.green * game.dice.blue, 0);
  console.log('part 2', power);
};

if (PART1) part1();
if (PART2) part2();
